symfit/core/fit.py

```diff
@@ -238,32 +356,59 @@ class BaseNumericalModel(BaseModel):
     ABC for Numerical Models. These are models whose components are generic
     python callables.
     """
-    def __init__(self, model, independent_vars, params):
+    @keywordonly(connectivity_mapping=None)
+    def __init__(self, model, independent_vars=None, params=None, **kwargs):
         """
         :param model: dict of ``callable``, where dependent variables are the
             keys. If instead of a dict a (sequence of) ``callable`` is provided,
             it will be turned into a dict automatically.
-        :param independent_vars: The independent variables of the model.
+        :param independent_vars: The independent variables of the  model.
+            (Deprecated, use ``connectivity_mapping`` instead.)
         :param params: The parameters of the model.
-        """
-        self.independent_vars = sorted(independent_vars, key=str)
-        self.params = sorted(params, key=str)
+            (Deprecated, use ``connectivity_mapping`` instead.)
+        :param connectivity_mapping: Mapping indicating the dependencies of
+            every variable in the model. For example, a model_dict
+            {y: a * x + b} has a connectivity_mapping {y: {x, a, b}}. Note that
+            the values of this dict have to be sets.
+        """
+        connectivity_mapping = kwargs.pop('connectivity_mapping')
+        if connectivity_mapping is None and \
+                independent_vars is not None and params is not None:
+            # Make model into a mapping if needed.
+            if not isinstance(model, Mapping):
+                try:
+                    iter(model)
+                except TypeError:
+                    model = [model]  # make model iterable
+
+                with warnings.catch_warnings():
+                    warnings.simplefilter("ignore")
+                    model = {Variable(): expr for expr in model}
+            warnings.warn(DeprecationWarning(
+                '`independent_vars` and `params` have been deprecated.'
+                ' Use `connectivity_mapping` instead.'
+            ))
+            self.independent_vars = sorted(independent_vars, key=str)
+            self.params = sorted(params, key=str)
+            self.connectivity_mapping = {var: set(independent_vars + params)
+                                         for var in model}
```

Comment:
Everything depends on everything? Can't we do better?

Reply:
I don't think that we can if we don't want to use inspect to much. Without using `inspect` on the provided functions we cannot know which kwargs to provide.

We could perhaps check for the existence of an `__signature__` attribute, but that is not present on most callable by default to my knowledge, unless explicitly set.

The good news is, that making everything a function of everything was the convention up to this point, so we now just do better if we can.